### PR TITLE
3 d back simple

### PR DESCRIPTION
**Pull request by vcloarec**
## Checklist

- [ ] Commit messages are descriptive and explain the rationale for changes
- [ ] Commits which fix bugs include `Fixes #11111` at the bottom of the commit message
- [ ] I have read the [QGIS Coding Standards](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
- [ ] New unit tests have been added for core changes
- [ ] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTING.md#contributing-to-qgis) before each commit
- [ ] I have evaluated whether it is appropriate for this PR to be backported, backport requests are left as label or comment

### CODE REVIEW

File: src/3d/shaders/mesh/mesh.frag [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+
+void main()
+{
+    bool wireframe=true;
+    float ambianceFactor=0.15;
+
+    vec3 diffuseColor;
+    vec3 specularColor;
+    vec4 color=vec4(0.,0.0,1.0,1);
+
```
Comment (vcloarec) [inline src/3d/shaders/mesh/mesh.frag]: comment the light applying or the wireframe applying resolve the issue of random moving triangle.
I think that comment one or other works because, in openglsl, if a uniform is not used in the code, then the compiler doesn't take it into account at all.

File: src/3d/symbols/qgsmesh3dsymbol.h [hunk 2]
```diff
@@ -36,6 +37,33 @@
...
+     *
+     * \since QGIS 3.12
+     */
+    enum RendererType
+    {
+      //! Render the mesh as a simple 3D symbol
+      simpleSymbology = 0,
+      //! Render the mesh as advanced symbology (smoothed triangled, wireframe, contours...)
+      advancedSymbology
+    };
```
Comment (wonder-sk) [inline src/3d/symbols/qgsmesh3dsymbol.h]: this can be probably removed?

File: src/3d/symbols/qgsmesh3dsymbol.h [hunk 3]
```diff
@@ -36,6 +37,33 @@
...
+     *
+     * \since QGIS 3.12
+     */
+    enum TextureType
+    {
+      //! Render the mesh with a unique color
+      uniqueColor = 0,
+      //! Render the mesh with a color ramp
+      colorRamp
+    };
```
Comment (wonder-sk) [inline src/3d/symbols/qgsmesh3dsymbol.h:52]: Can we have upper case starting letters?

And maybe rename TextureType to something else? Usage of texture is just an implementation detail... How about "RenderingStyle" ?

File: src/3d/terrain/qgsterraingenerator.h [hunk 4]
```diff
@@ -49,6 +49,7 @@ class _3D_EXPORT QgsTerrainGenerator : public QgsChunkLoaderFactory
       Flat,           //!< The whole terrain is flat area
       Dem,            //!< Terrain is built from raster layer with digital elevation model
       Online,         //!< Terrain is built from downloaded tiles with digital elevation model
+      Mesh
```
Comment (wonder-sk) [inline src/3d/terrain/qgsterraingenerator.h]: add doc please

File: src/app/3d/qgs3dmapcanvas.cpp [hunk 5]
```diff
@@ -42,6 +42,7 @@ Qgs3DMapCanvas::Qgs3DMapCanvas( QWidget *parent )
   } );
 
   mContainer = QWidget::createWindowContainer( mEngine->window() );
+  mContainer->installEventFilter( this );
```
Comment (wonder-sk) [inline src/app/3d/qgs3dmapcanvas.cpp]: why is this necessary?

File: src/app/3d/qgs3dmapcanvas.cpp [hunk 6]
```diff
@@ -163,25 +163,40 @@ void Qgs3DMapCanvas::setMapTool( Qgs3DMapTool *tool )
...
-    case QEvent::MouseButtonRelease:
-      mMapTool->mouseReleaseEvent( static_cast<QMouseEvent *>( event ) );
-      break;
-    case QEvent::MouseMove:
-      mMapTool->mouseMoveEvent( static_cast<QMouseEvent *>( event ) );
-      break;
-    default:
-      break;
+    if ( event->type() == QEvent::Resize )
+      updateViewportSize();
```
Comment (wonder-sk) [inline src/app/3d/qgs3dmapcanvas.cpp]: Are the changes in this file still necessary? Or can we roll them back?

File: src/3d/mesh/qgsmesh3dgeometry_p.cpp [hunk 7]
```diff
@@ -0,0 +1,421 @@
...
+
+  mIndexAttribute->setAttributeType( QAttribute::IndexAttribute );
+  mIndexAttribute->setVertexBaseType( QAttribute::UnsignedInt );
+  mIndexAttribute->setBuffer( mIndexBuffer );
+
+  // Each primitive has 3 vertives
+  mIndexAttribute->setCount( uint( mTriangularMesh.triangles().count() ) * 3 );
+
+  mVertexBuffer->setData( MeshPlaneVertexBufferFunctor( mTriangularMesh, mExtent, mVertScale )() );
+  mIndexBuffer->setData( MeshPlaneIndexBufferFunctor( mTriangularMesh )() );
```
Comment (wonder-sk) [inline src/3d/mesh/qgsmesh3dgeometry_p.cpp:317]: I would suggest to remove the functors and simply call the create*() functions directly here - to simplify the code

File: src/3d/mesh/qgsmesh3dgeometry_p.cpp [hunk 8]
```diff
@@ -0,0 +1,421 @@
...
+  private:
+    QgsTriangularMesh mMesh;
+    QVector<double> mVerticaleMagnitude;
+    QVector<double> mScalarMagnitude;
+    QgsRectangle mExtent;
+    float mVertScale;
+
+};
+
+QgsMesh3dGeometry_p::QgsMesh3dGeometry_p(
```
Comment (wonder-sk) [inline src/3d/mesh/qgsmesh3dgeometry_p.cpp]: Can we remove the _p suffix from the class name? We only use the _p suffix for file names of files that are part of the private implementations, but not for class names...

File: src/3d/mesh/qgsmesh3dmaterial.h [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+
+
+class QgsMesh3dMaterial : public Qt3DRender::QMaterial
+{
+  public:
+    enum MagnitudeType
+    {
+      zValue,
+      scalarDataSet
+    };
```
Comment (wonder-sk) [inline src/3d/mesh/qgsmesh3dmaterial.h]: Can we have the values starting with upper case? And document them?

File: src/3d/qgs3dmapscene.cpp [hunk 10]
```diff
@@ -309,8 +314,11 @@ bool Qgs3DMapScene::updateCameraNearFarPlanes()
           ffar = dst;
       }
     }
-    if ( fnear < 1 )
-      fnear = 1;  // does not really make sense to use negative far plane (behind camera)
+
+    float fnearMini = 1;
+
+    if ( fnear < fnearMini )
+      fnear = fnearMini;
```
Comment (wonder-sk) [inline src/3d/qgs3dmapscene.cpp]: It looks like we have lost the comment here explaining why it is 1

File: src/3d/qgs3dmapscene.cpp [hunk 11]
```diff
@@ -222,6 +222,11 @@ float Qgs3DMapScene::worldSpaceError( float epsilon, float distance )
   return err;
 }
 
+void Qgs3DMapScene::setViewport( const QRect &viewportrect )
+{
+  cameraController()->setViewport( viewportrect );
+}
+
```
Comment (wonder-sk) [inline src/3d/qgs3dmapscene.cpp]: Is this new method still needed? Better to remove it if not

File: src/3d/qgs3dmapsettings.h [hunk 12]
```diff
@@ -446,6 +459,7 @@ class _3D_EXPORT Qgs3DMapSettings : public QObject
     QgsPathResolver mPathResolver;
     QgsMapThemeCollection *mMapThemes = nullptr;   //!< Pointer to map themes (e.g. from the current project) to resolve map theme content from the name
     double mDpi = 96;  //!< Dot per inch value for the screen / painter
+    QgsMesh3DSymbol mMeshterrainSymbol;
```
Comment (wonder-sk) [inline src/3d/qgs3dmapsettings.h]: Why is mesh terrain symbol defined here? I think it should be defined inside the mesh terrain generator...

File: src/3d/shaders/mesh/mesh.frag [hunk 13]
```diff
@@ -0,0 +1,275 @@
...
+        float x = d - ( lineWidth - 1.0 );
+        mixVal = exp2( -2.0 * ( x * x ) );
+    }
+
+    return mix( color, lineColor, mixVal );
+}
+
+void adModel(const in vec3 worldPos,
+             const in vec3 worldNormal,
+             out vec3 diffuseColor)
```
Comment (wonder-sk) [inline src/3d/shaders/mesh/mesh.frag]: Is it possible to reuse the existing adsModel implementation that exists in light.inc.frag? or is it not possible because of something? It is quite a lot of duplicated code...

File: src/3d/shaders/mesh/mesh.frag [hunk 14]
```diff
@@ -0,0 +1,275 @@
...
+    float constantAttenuation;
+    float linearAttenuation;
+    float quadraticAttenuation;
+    float cutOffAngle;
+};
+uniform Light lights[MAX_LIGHTS];
+uniform int lightCount;
+
+
+vec4 wireframeShadeLine( const in vec4 color )
```
Comment (wonder-sk) [inline src/3d/shaders/mesh/mesh.frag:29]: it would be good to point to the original code for reference

File: src/3d/shaders/mesh/mesh.geom [hunk 15]
```diff
@@ -0,0 +1,160 @@
+#version 330 core
```
Comment (wonder-sk) [inline src/3d/shaders/mesh/mesh.geom:1]: it would be good to point to the original code of this file for reference

File: src/3d/shaders/mesh/mesh.geom [hunk 16]
```diff
@@ -0,0 +1,160 @@
...
+
+layout( triangles ) in;
+layout( triangle_strip, max_vertices = 3 ) out;
+
+uniform mat4 viewportMatrix;
+uniform bool flatTriangles;
+
+in MeshVertex {
+    vec3 wordPosition;
```
Comment (wonder-sk) [inline src/3d/shaders/mesh/mesh.geom]: shouldn't this be `world` instead of `word` ?

File: src/3d/shaders/mesh/mesh.vert [hunk 17]
```diff
@@ -0,0 +1,27 @@
...
+
+out MeshVertex {
+    vec3 wordPosition;
+    vec3 wordNormal;
+    float magnitude;
+} vs_out;
+
+void main()
+{
+    //vs_out.position = vec3( modelView * vec4( vertexPosition, 1.0 ) );
```
Comment (wonder-sk) [inline src/3d/shaders/mesh/mesh.vert]: best to remove commented out code ...

File: src/core/mesh/qgstriangularmesh.cpp [hunk 18]
```diff
@@ -92,6 +92,28 @@ void QgsTriangularMesh::triangulate( const QgsMeshFace &face, int nativeIndex )
   mTrianglesToNativeFaces.push_back( nativeIndex );
 }
 
+
+void QgsTriangularMesh::setCCW( QgsMeshFace &face )
```
Comment (wonder-sk) [inline src/core/mesh/qgstriangularmesh.cpp]: better to expand CCW so it is clear for everyone...

File: src/core/mesh/qgstriangularmesh.cpp [hunk 19]
```diff
@@ -223,6 +251,31 @@ QList<int> QgsTriangularMesh::faceIndexesForRectangle( const QgsRectangle &recta
   return mSpatialIndex.intersects( rectangle );
 }
 
+QVector<QVector3D> QgsTriangularMesh::vertexNormales( float vertScale ) const
```
Comment (wonder-sk) [inline src/core/mesh/qgstriangularmesh.cpp]: Normales -> Normals

File: src/core/mesh/qgstriangularmesh.cpp [hunk 20]
```diff
@@ -223,6 +251,31 @@ QList<int> QgsTriangularMesh::faceIndexesForRectangle( const QgsRectangle &recta
...
+{
+  QVector<QVector3D> normales( vertices().count(), QVector3D( 0, 0, 0 ) );
+
+  for ( const auto &face : triangles() )
+  {
+    for ( int i = 0; i < 3; i++ )
+    {
+      int index1( face.at( i ) );
+      int index2( face.at( ( i + 1 ) % 3 ) );
+      int index3( face.at( ( i + 2 ) % 3 ) );
```
Comment (wonder-sk) [inline src/core/mesh/qgstriangularmesh.cpp:264]: why % 3 ?

Comment (vcloarec) [reply src/core/mesh/qgstriangularmesh.cpp:264]: if i = 1 or  to access  to the index 0 with i+2
if i = 2 to access to the index 0 and 1 with i+1 and i+2

File: src/3d/mesh/qgsmesh3dgeometry_p.cpp (hunk 7 above)
Comment (vcloarec) [reply src/3d/mesh/qgsmesh3dgeometry_p.cpp:317]: This the way used in the example of Qt3d, and the used for `DemTerrainTileGeometry` ... but you are right, it would simplify the code. Or maybe, it is better to use setDataGenerator(), so the work could done asynchronously (https://doc.qt.io/qt-5/qt3drender-qbuffer.html#details)

File: src/3d/qgs3dmapsettings.h (hunk 12 above)
Comment (vcloarec) [reply src/3d/qgs3dmapsettings.h]: If this symbol is defined in the mesh terrain generator, it would be destruct if the terrain generator is changed. Here, it could be saved event if the terrain is changed.

File: src/3d/shaders/mesh/mesh.frag (hunk 14 above)
Comment (vcloarec) [reply src/3d/shaders/mesh/mesh.frag:29]: sure!

File: src/3d/shaders/mesh/mesh.frag (hunk 13 above)
Comment (vcloarec) [reply src/3d/shaders/mesh/mesh.frag]: I will try, but i have duplicate code to have more simple shader (not the apparent one, but the compiled one) without other light fucntion and other stuff we don't need.
And I wonder, if the issue didn't come from that include...
I will try.

File: src/3d/shaders/mesh/mesh.geom (hunk 15 above)
Comment (vcloarec) [reply src/3d/shaders/mesh/mesh.geom:1]: sure!

File: src/3d/symbols/qgsmesh3dsymbol.h (hunk 2 above)
Comment (vcloarec) [reply src/3d/symbols/qgsmesh3dsymbol.h]: sure!

File: src/app/3d/qgs3dmapcanvas.cpp (hunk 5 above)
Comment (vcloarec) [reply src/app/3d/qgs3dmapcanvas.cpp]: not anymore...

File: src/app/3d/qgs3dmapcanvas.cpp (hunk 6 above)
Comment (vcloarec) [reply src/app/3d/qgs3dmapcanvas.cpp]: sure!

File: src/3d/shaders/mesh/mesh.frag (hunk 13 above)
Comment (vcloarec) [reply src/3d/shaders/mesh/mesh.frag]: Finaly, it is ok, so ok...

File: src/3d/shaders/mesh/mesh.geom (hunk 16 above)
Comment (vcloarec) [reply src/3d/shaders/mesh/mesh.geom]: of course...

File: src/3d/mesh/qgsmesh3dgeometry_p.cpp (hunk 7 above)
Comment (wonder-sk) [reply src/3d/mesh/qgsmesh3dgeometry_p.cpp:317]: I am quite easy with this - feel free to also skip this point...

File: src/3d/qgs3dmapsettings.h (hunk 12 above)
Comment (wonder-sk) [reply src/3d/qgs3dmapsettings.h]: I think it is fine if the symbol gets "lost" when changing to a different terrain generator. If the symbol would be inside the mesh terrain generator, it would be cleaner solution.

File: src/core/mesh/qgstriangularmesh.cpp (hunk 20 above)
Comment (wonder-sk) [reply src/core/mesh/qgstriangularmesh.cpp:264]: of course - sorry I have misread the code...
